File: imageCropper/src/main/java/com/crust87/imagecropper/ImageCropper.java
/*
 * ImageCropper
 * https://github.com/crust87/Android-ImageCropper
 *
 * Mabi
 * crust87@gmail.com
 * last modify 2015-05-21
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.crust87.imagecropper;

import android.content.Context;
import android.content.res.TypedArray;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.Rect;
import android.media.ExifInterface;
import android.net.Uri;
import android.provider.MediaStore;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.SurfaceHolder;
import android.view.SurfaceView;

import com.crust87.imagecropper.cropbox.CropBox;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;

public class ImageCropper extends SurfaceView implements SurfaceHolder.Callback {
	// Constants
	public static final int CIRCLE_CROP_BOX = 0;
	public static final int RECT_CROP_BOX = 1;

	private static final int DEFAULT_BOX_TYPE = CIRCLE_CROP_BOX;

	// Components
	private Context mContext;
	private SurfaceHolder mHolder;
	private Bitmap mImage;
	private Bitmap mScaledImage;
	private Rect mImageBound;
	private Paint mImagePaint;
	private CropBox.CropBoxBuilder mCropBoxBuilder;
	private CropBox mCropBox;

	// Attributes
	private int mViewWidth;
	private int mViewHeight;
	private String mImagePath;
	private boolean isImageOpen;

	// Listener
	private OnCropBoxChangedListener mOnCropBoxChangedListener;

	// Constructor
	public ImageCropper(Context context) {
		super(context);
		mContext = context;
		mCropBoxBuilder = new CropBox.CropBoxBuilder();

		initAttributes();
		initImageCropper();
	}

	public ImageCropper(Context context, AttributeSet attrs) {
		super(context, attrs);
		mContext = context;
		mCropBoxBuilder = new CropBox.CropBoxBuilder();

		initAttributes(context, attrs, 0);
		initImageCropper();
	}

	public ImageCropper(Context context, AttributeSet attrs, int defStyleAttr) {
		super(context, attrs, defStyleAttr);
		mContext = context;
		mCropBoxBuilder = new CropBox.CropBoxBuilder();

		initAttributes(context, attrs, defStyleAttr);
		initImageCropper();
	}

	private void initAttributes() {
		mCropBoxBuilder.setBoxColor(getResources().getColor(R.color.default_box_color))
				.setBoxType(DEFAULT_BOX_TYPE)
				.setLineWidth(getResources().getDimensionPixelSize(R.dimen.default_line_width))
				.setAnchorSize(getResources().getDimensionPixelSize(R.dimen.default_anchor_size));
	}

	private void initAttributes(Context context, AttributeSet attrs, int defStyleAttr) {
		TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.ImageCropper, defStyleAttr, 0);

		int boxColor = typedArray.getColor(R.styleable.ImageCropper_box_color, getResources().getColor(R.color.default_box_color));
		int boxType = CIRCLE_CROP_BOX;

		String lBoxType = typedArray.getString(R.styleable.ImageCropper_box_type);
		if(TextUtils.equals(lBoxType, "circle")) {
			boxType = CIRCLE_CROP_BOX;
		} else if(TextUtils.equals(lBoxType, "rect")) {
			boxType = RECT_CROP_BOX;
		}

		int defaultLineWidth = getResources().getDimensionPixelSize(R.dimen.default_line_width);
		int defaultAnchorSize = getResources().getDimensionPixelSize(R.dimen.default_anchor_size);
		int lineWidth = typedArray.getLayoutDimension(R.styleable.ImageCropper_line_width, defaultLineWidth);
		int anchorSize = typedArray.getLayoutDimension(R.styleable.ImageCropper_anchor_size, defaultAnchorSize);

		mCropBoxBuilder.setBoxColor(boxColor)
				.setBoxType(boxType)
				.setLineWidth(lineWidth)
				.setAnchorSize(anchorSize);
	}

	private void initImageCropper() {
		mHolder = getHolder();
		mHolder.addCallback(this);
		isImageOpen = false;
	}

	/**
	 *
	 * @param imageUri it will convert into real path
     */
	public void setImage(Uri imageUri) {
		setImage(getRealPathFromURI(imageUri));
	}

	/**
	 *box_color
	 * @param imageFile it will convert into absolute path
     */
	public void setImage(File imageFile) {
		setImage(imageFile.getAbsolutePath());
	}

	/**
	 *
	 * @param imagePath path of image to draw on view
     */
	public void setImage(String imagePath) {
		mImagePath = imagePath;

		openImage();
		invalidate();
	}

	@Override
	public void surfaceCreated(SurfaceHolder holder) {
		setWillNotDraw(false);
		mViewWidth = getWidth();
		mViewHeight = getHeight();

		openImage();
	}

	@Override
	public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
		mViewWidth = width;
		mViewHeight = height;
	}

	@Override
	public void surfaceDestroyed(SurfaceHolder holder) {
		destroyDrawingCache();
	}

	@Override
	public boolean onTouchEvent(MotionEvent event) {
		if(mCropBox != null) {
			boolean isChanged = mCropBox.processTouchEvent(event);

			if(isChanged) {
				if(mOnCropBoxChangedListener != null) {
					mOnCropBoxChangedListener.onCropBoxChange(mCropBox);
				}
			}

			invalidate();
		}

		return true;
	}

	private void openImage() {
		if(mViewWidth == 0 || mViewHeight == 0) {
			return;
		}

		if(TextUtils.isEmpty(mImagePath)) {
			return;
		}

		InputStream imageStream = null;

		try {
			imageStream = new FileInputStream(mImagePath);
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		}

		// check rotation
		int lRotaion = 0;
		try {
			ExifInterface exif = new ExifInterface(mImagePath);

			switch(Integer.parseInt(exif.getAttribute(ExifInterface.TAG_ORIENTATION))) {
				case 3:
					lRotaion = 180;
					break;
				case 6:
					lRotaion = 90;
					break;
				case 8:
					lRotaion = 270;
					break;
				default:
					lRotaion = 0;
			}
		} catch (IllegalArgumentException e) {
			e.printStackTrace();
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}

		BitmapFactory.Options options = new BitmapFactory.Options();
		options.inPreferredConfig = Bitmap.Config.RGB_565;

		Matrix matrix = new Matrix();
		matrix.postRotate(lRotaion);

		mImage = BitmapFactory.decodeStream(imageStream, null, options);
		mImage = Bitmap.createBitmap(mImage, 0, 0, mImage.getWidth(), mImage.getHeight(), matrix, true);

		mImagePaint = new Paint();
		mImagePaint.setFilterBitmap(true);

		float scaleX;
		float scaleY;

		scaleX = (float) getWidth() / mImage.getWidth();
		scaleY = (float) getHeight() / mImage.getHeight();

		int newWidth;
		int newHeight;
		float lScale;

		if(scaleX > scaleY) {
			lScale = scaleY;
		} else {
			lScale = scaleX;
		}

		newWidth = (int) (mImage.getWidth() * lScale);
		newHeight = (int) (mImage.getHeight() * lScale);

		mScaledImage = Bitmap.createScaledBitmap(mImage, newWidth, newHeight, true);

		int lLeftMargin = (mViewWidth - mScaledImage.getWidth()) / 2;
		int lTopMargin = (mViewHeight - mScaledImage.getHeight()) / 2;

		mImageBound = new Rect(lLeftMargin, lTopMargin, mScaledImage.getWidth() + lLeftMargin, mScaledImage.getHeight() + lTopMargin);

		mCropBoxBuilder.setLeftMargin(lLeftMargin)
				.setTopMargin(lTopMargin)
				.setBound(mImageBound)
				.setScale(lScale);

		mCropBox = mCropBoxBuilder.createCropBox(getContext());

		if(mOnCropBoxChangedListener != null) {
			mOnCropBoxChangedListener.onCropBoxChange(mCropBox);
		}

		isImageOpen = true;
	}

	protected void onDraw(Canvas canvas) {
		super.onDraw(canvas);
		
		drawPicture(canvas);
		drawCropBox(canvas);
	}
	
	private void drawCropBox(Canvas pCanvas) {
		if(mCropBox != null) {
			mCropBox.draw(pCanvas);
		}
	}
	
	private void drawPicture(Canvas pCanvas) {
		if(mScaledImage != null) {
			pCanvas.drawBitmap(mScaledImage, null, mImageBound, mImagePaint);
		}
	}

	/**
	 * crop image
	 *
	 * @return Bitmap of cropped image
     */
	public Bitmap crop() {
		if(!isImageOpen) {
			return null;
		}

		int lCropX = mCropBox.getCropX();
		int lCropY = mCropBox.getCropY();
		int lCropWidth = mCropBox.getCropWidth();
		int lCropHeight = mCropBox.getCropHeight();

		if(lCropX + lCropWidth > mImage.getWidth()) {
			lCropWidth -= lCropX + lCropWidth - mImage.getWidth();
		}

		if(lCropY + lCropHeight > mImage.getHeight()) {
			lCropHeight -= lCropY + lCropHeight - mImage.getHeight();
		}

		return Bitmap.createBitmap(mImage, lCropX, lCropY, lCropWidth, lCropHeight);
	}

	public Bitmap getOriginalImage() {
		if(!isImageOpen) {
			return null;
		}

		return mImage;
	}

	private String getRealPathFromURI(Uri contentUri) {
		Cursor cursor = null;
		try {
			String[] proj = { MediaStore.Images.Media.DATA };
			cursor = mContext.getContentResolver().query(contentUri, proj, null, null, null);
			int column_index = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATA);
			cursor.moveToFirst();
			return cursor.getString(column_index);
		} finally {
			if (cursor != null) {
				cursor.close();
			}
		}
	}

	public void setOnCropBoxChangedListener(OnCropBoxChangedListener pOnCropBoxChangedListener) {
		mOnCropBoxChangedListener = pOnCropBoxChangedListener;
	}

	public interface OnCropBoxChangedListener {
		public abstract void onCropBoxChange(CropBox cropBox);
	}

	public void setBoxColor(int color) {
		mCropBoxBuilder.setBoxColor(color);

		if(mCropBox != null) {
			mCropBox.setColor(color);
		}
	}

	public void setBoxColor(String colorCode) {
		setBoxColor(Color.parseColor(colorCode));
	}

	public void setBoxType(int boxType) {
		mCropBoxBuilder.setBoxType(boxType);

		if(isImageOpen) {
			mCropBox = mCropBoxBuilder.createCropBox(getContext());
			invalidate();
		}
	}

	public void setLineWidth(int lineWidth) {
		mCropBoxBuilder.setLineWidth(lineWidth);

		if(mCropBox != null) {
			mCropBox.setLineWidth(lineWidth);
		}
	}

	public void setAnchorSize(int anchorSize) {
		mCropBoxBuilder.setAnchorSize(anchorSize);

		if(mCropBox != null) {
			mCropBox.setAnchorSize(anchorSize);
		}
	}

	public boolean isImageOpen() {
		return isImageOpen;
	}
}

File: library/src/main/java/com/crust87/imagecropper/ImageCropper.kt
/*
 * ImageCropper
 * https://github.com/crust87/Android-ImageCropper
 *
 * Mabi
 * crust87@gmail.com
 * last modify 2015-05-21
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.crust87.imagecropper

import android.content.Context
import android.database.Cursor
import android.graphics.*
import android.media.ExifInterface
import android.net.Uri
import android.provider.MediaStore
import android.text.TextUtils
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.SurfaceHolder
import android.view.SurfaceView
import android.view.ViewConfiguration
import com.crust87.imagecropper.cropbox.CropBox
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.IOException

class ImageCropper : SurfaceView, SurfaceHolder.Callback {

    private val cropBoxBuilder: CropBox.CropBoxBuilder = CropBox.CropBoxBuilder()
    private var originImage: Bitmap? = null
    private var displayImage: Bitmap? = null
    private val imageBound = RectF()
    private var imagePaint = Paint().apply {
        isFilterBitmap = true
    }
    private var cropBox: CropBox? = null

    /*
    Attributes
     */
    private var viewWidth = 0
    private var viewHeight = 0
    private var imagePath: String? = null
    private var scale = 0f

    var isOpened: Boolean = false
        private set

    private var onCropBoxChangedListener: OnCropBoxChangedListener? = null

    /*
    Constructors
     */
    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(context, attrs, 0)
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init(context, attrs, defStyleAttr)
    }

    private fun init() {
        with(cropBoxBuilder) {
            minSize = context.resources.getDimensionPixelSize(R.dimen.min_box_Size).toFloat()
            touchSlop = ViewConfiguration.get(context).scaledTouchSlop.toFloat()
            boxColor = resources.getColor(R.color.default_box_color)
            lineWidth = resources.getDimensionPixelSize(R.dimen.default_line_width)
            anchorSize = resources.getDimensionPixelSize(R.dimen.default_anchor_size)
        }

        holder.addCallback(this)
        isOpened = false
    }

    private fun init(context: Context, attrs: AttributeSet, defStyleAttr: Int) {
        val typedArray = context.obtainStyledAttributes(attrs, R.styleable.ImageCropper, defStyleAttr, 0)

        val boxColor = typedArray.getColor(R.styleable.ImageCropper_box_color, resources.getColor(R.color.default_box_color))
        val boxType = when (typedArray.getString(R.styleable.ImageCropper_box_type)) {
            "circle" -> CIRCLE_CROP_BOX
            "rect" -> RECT_CROP_BOX
            else -> CIRCLE_CROP_BOX
        }
        val defaultLineWidth = resources.getDimensionPixelSize(R.dimen.default_line_width)
        val defaultAnchorSize = resources.getDimensionPixelSize(R.dimen.default_anchor_size)
        val lineWidth = typedArray.getLayoutDimension(R.styleable.ImageCropper_line_width, defaultLineWidth)
        val anchorSize = typedArray.getLayoutDimension(R.styleable.ImageCropper_anchor_size, defaultAnchorSize)

        with(cropBoxBuilder) {
            this.boxType = boxType
            this.minSize = context.resources.getDimensionPixelSize(R.dimen.min_box_Size).toFloat()
            this.touchSlop = ViewConfiguration.get(context).scaledTouchSlop.toFloat()
            this.boxColor = boxColor
            this.lineWidth = lineWidth
            this.anchorSize = anchorSize
        }

        holder.addCallback(this)
        isOpened = false

        typedArray.recycle()
    }

    /**
     *
     * @param uri it will convert into real path
     */
    fun setImage(uri: Uri) {
        setImage(getRealPathFromURI(uri))
    }

    /**
     * box_color
     * @param file it will convert into absolute path
     */
    fun setImage(file: File) {
        setImage(file.absolutePath)
    }

    /**
     *
     * @param path path of image to draw on view
     */
    fun setImage(path: String) {
        imagePath = path

        openImage()
        invalidate()
    }

    override fun surfaceCreated(holder: SurfaceHolder) {
        setWillNotDraw(false)
        viewWidth = width
        viewHeight = height

        openImage()
    }

    override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
        viewWidth = width
        viewHeight = height
    }

    override fun surfaceDestroyed(holder: SurfaceHolder) {
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        cropBox?.let { cropBox ->
            val isChanged = cropBox.processTouchEvent(event)

            if (isChanged) {
                onCropBoxChangedListener?.onCropBoxChange(cropBox)
            }

            invalidate()
        }

        return true
    }

    private fun openImage() {
        if (viewWidth == 0 || viewHeight == 0) {
            return
        }

        if (TextUtils.isEmpty(imagePath)) {
            return
        }

        try {
            val imageStream = FileInputStream(imagePath)
            val exif = ExifInterface(imagePath)
            val rotation = when (Integer.parseInt(exif.getAttribute(ExifInterface.TAG_ORIENTATION))) {
                3 -> 180F
                6 -> 90F
                8 -> 270F
                else -> 0F
            }

            val options = BitmapFactory.Options().apply {
                inPreferredConfig = Bitmap.Config.RGB_565
            }
            val decodedImage = BitmapFactory.decodeStream(imageStream, null, options)?.run {
                val matrix = Matrix().apply {
                    postRotate(rotation)
                }

                Bitmap.createBitmap(this, 0, 0, this.width, this.height, matrix, true)
            }

            decodedImage ?: return

            val scaleX = width.toFloat() / decodedImage.width
            val scaleY = height.toFloat() / decodedImage.height

            scale = when (scaleX > scaleY) {
                true -> scaleY
                false -> scaleX
            }
            val newWidth = (decodedImage.width * scale).toInt()
            val newHeight = (decodedImage.height * scale).toInt()

            val scaledImage = Bitmap.createScaledBitmap(decodedImage, newWidth, newHeight, true)

            scaledImage ?: return

            val leftMargin = (viewWidth - scaledImage.width) / 2f
            val topMargin = (viewHeight - scaledImage.height) / 2f

            imageBound.left = leftMargin
            imageBound.top = topMargin
            imageBound.right = scaledImage.width + leftMargin
            imageBound.bottom = scaledImage.height + topMargin

            with(cropBoxBuilder) {
                this.bound = imageBound
            }

            cropBox = cropBoxBuilder.createCropBox().apply {
                onCropBoxChangedListener?.onCropBoxChange(this)
            }

            originImage = decodedImage
            displayImage = scaledImage
            isOpened = true
        } catch (e: IllegalArgumentException) {
            e.printStackTrace()
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        drawPicture(canvas)
        drawCropBox(canvas)
    }

    private fun drawCropBox(canvas: Canvas) {
        cropBox?.draw(canvas)
    }

    private fun drawPicture(canvas: Canvas) {
        displayImage?.let { displayImage ->
            canvas.drawBitmap(displayImage, null, imageBound, imagePaint)
        }
    }

    /**
     * crop image
     *
     * @return Bitmap of cropped image
     */
    fun crop(): Bitmap? {
        if (!isOpened) {
            return null
        }

        val originImage = originImage
        val cropBox = cropBox

        originImage ?: return null
        cropBox ?: return null


        val x = (cropBox.x / scale).toInt()
        val y = (cropBox.y / scale).toInt()
        var width = (cropBox.width / scale).toInt()
        var height = (cropBox.height / scale).toInt()

        if (x + width > originImage.width) {
            width -= x + width - originImage.width
        }

        if (y + height > originImage.height) {
            height -= y + height - originImage.height
        }

        return Bitmap.createBitmap(originImage, x, y, width, height)
    }

    fun setOnCropBoxChangedListener(l: OnCropBoxChangedListener) {
        onCropBoxChangedListener = l
    }

    fun setOnCropBoxChangedListener(l: (CropBox) -> Unit) {
        onCropBoxChangedListener = object : OnCropBoxChangedListener {
            override fun onCropBoxChange(cropBox: CropBox) {
                l(cropBox)
            }
        }
    }

    interface OnCropBoxChangedListener {
        fun onCropBoxChange(cropBox: CropBox)
    }

    fun setBoxType(boxType: Int) {
        cropBoxBuilder.boxType = boxType

        if (isOpened) {
            cropBox = cropBoxBuilder.createCropBox()
            invalidate()
        }
    }

    private fun getRealPathFromURI(contentUri: Uri): String {
        var cursor: Cursor? = null
        try {
            val projection = arrayOf(MediaStore.Images.Media.DATA)
            cursor = context.contentResolver.query(contentUri, projection, null, null, null)
            val columnIndex = cursor!!.getColumnIndexOrThrow(MediaStore.Images.Media.DATA)
            cursor.moveToFirst()
            return cursor.getString(columnIndex)
        } finally {
            cursor?.close()
        }
    }

    companion object {
        const val CIRCLE_CROP_BOX = 0
        const val RECT_CROP_BOX = 1
        const val DEFAULT_BOX_TYPE = CIRCLE_CROP_BOX
    }
}
